pkg/egressgateway/manager.go

```diff
@@ -266,15 +266,98 @@ func (manager *Manager) updatePoliciesBySourceIP() {
+				ip := epIP.String()
+				manager.policyConfigsBySourceIP[ip] = append(manager.policyConfigsBySourceIP[ip], policy)
+			}
+		}
+	}
+}
+
+// policyMatches returns true if there exists at least one policy matching the
+// given parameters.
+//
+// This method takes:
+//   - a source IP: this is an optimization that allows to iterate only through
+//     policies that reference an endpoint with the given source IP
+//   - a callback function f: this function is invoked for each policy and for
+//     each combination of the policy's endpoints and destination/excludedCIDRs.
+//
+// The callback f takes as arguments:
+// - the given endpoint
+// - the destination CIDR
+// - a boolean value indicating if the CIDR belongs to the excluded ones
+// - the gatewayConfig of the  policy
+//
+// This method returns true whenever the f callback matches one of the endpoint
+// and CIDR tuples (i.e. whenever one callback invocation returns true)
+func (manager *Manager) policyMatches(sourceIP net.IP, f func(net.IP, *net.IPNet, bool, *gatewayConfig) bool) bool {
+	for _, policy := range manager.policyConfigsBySourceIP[sourceIP.String()] {
+		for _, ep := range policy.matchedEndpoints {
+			for _, endpointIP := range ep.ips {
+				if !endpointIP.Equal(sourceIP) {
+					continue
+				}
+
+				isExcludedCIDR := false
+				for _, dstCIDR := range policy.dstCIDRs {
+					if f(endpointIP, dstCIDR, isExcludedCIDR, &policy.gatewayConfig) {
+						return true
+					}
+				}
+
+				isExcludedCIDR = true
+				for _, excludedCIDR := range policy.excludedCIDRs {
+					if f(endpointIP, excludedCIDR, isExcludedCIDR, &policy.gatewayConfig) {
+						return true
+					}
+				}
+			}
+		}
+	}
+
+	return false
+}
+
+// policyMatchesMinusExcludedCIDRs returns true if there exists at least one
+// policy matching the given parameters.
+//
+// This method takes:
+//   - a source IP: this is an optimization that allows to iterate only through
+//     policies that reference an endpoint with the given source IP
+//   - a callback function f: this function is invoked for each policy and for
+//     each combination of the policy's endpoints and computed destinations (i.e.
+//     the effective destination CIDR space, defined as the diff between the
+//     destination and the excluded CIDRs).
+//
+// The callback f takes as arguments:
+// - the given endpoint
+// - the destination CIDR
+// - the gatewayConfig of the  policy
+//
+// This method returns true whenever the f callback matches one of the endpoint
+// and CIDR tuples (i.e. whenever one callback invocation returns true)
+func (manager *Manager) policyMatchesMinusExcludedCIDRs(sourceIP net.IP, f func(net.IP, *net.IPNet, *gatewayConfig) bool) bool {
+	for _, policy := range manager.policyConfigsBySourceIP[sourceIP.String()] {
+		cidrs := policy.destinationMinusExcludedCIDRs()
+
+		for _, ep := range policy.matchedEndpoints {
+			for _, endpointIP := range ep.ips {
+				if !endpointIP.Equal(sourceIP) {
+					continue
+				}
```

Comment:
similar to above, if the `ep.ips` were sorted this could be an `O(log n)` binary search instead of linear scan